nightfall/detection_rules.py

```diff
@@ -0,0 +1,253 @@
+    is_case_sensitive: bool
+
+    def as_dict(self):
+        return {"pattern": self.pattern, "isCaseSensitive": self.is_case_sensitive}
+
+
+@dataclass
+class WordList:
+    """A list of words that can be used to customize the behavior of a detector while Nightfall performs a scan.
+    Attributes:
+        word_list (List[str]): The list of words to use.
+        is_case_sensitive (bool): Whether to make matches have the same case as each word given
+    """
+    word_list: List[str]
+    is_case_sensitive: bool
+
+    def as_dict(self):
+        return {"values": self.word_list, "isCaseSensitive": self.is_case_sensitive}
+
+
+class Confidence(Enum):
+    """Confidence describes the certainty that a piece of content matches a detector."""
+    VERY_UNLIKELY = "VERY_UNLIKELY"
+    UNLIKELY = "UNLIKELY"
+    POSSIBLE = "POSSIBLE"
+    LIKELY = "LIKELY"
+    VERY_LIKELY = "VERY_LIKELY"
+
+
+@dataclass
+class ContextRule:
+    """An object that describes how a regular expression may be used to adjust the confidence of a candidate finding.
+    This context rule will be applied within the provided byte proximity, and if the regular expression matches, then
+    the confidence associated with the finding will be adjusted to the value prescribed.
+    Attributes:
+        regex (Regex): The regular expression configuration to run within the context of a candidate finding.
+        window_before (int): The number of leading characters to consider as context.
+        window_after (int): The number of trailing characters to consider as context.
+        fixed_confidence (Confidence): How to adjust the result of the match if the context rule matches.
+    """
+    regex: Regex
+    window_before: int
+    window_after: int
+    fixed_confidence: Confidence
+
+    def as_dict(self):
+        return {
+            "regex": self.regex.as_dict(),
+            "proximity": {"windowBefore": self.window_before, "windowAfter": self.window_after},
+            "confidenceAdjustment": {"fixedConfidence": self.fixed_confidence.value}
+        }
+
+
+class MatchType(Enum):
+    FULL = "FULL"
+    PARTIAL = "PARTIAL"
+
+
+@dataclass
+class ExclusionRule:
+    """An object that describes a regular expression or list of keywords that may be used to disqualify a
+    candidate finding from triggering a detector match. One of regex or word_list is required.
+    Attributes:
+        match_type (MatchType): the match type.
+        regex (Regex or None): The regular expression configuration to run on a candidate finding.
+        word_list (WordList or None): The list of words to compare to a candidate finding.
+    """
+    match_type: MatchType
+    regex: Optional[Regex] = None
+    word_list: Optional[WordList] = None
+
+    def __post_init__(self):
+        if (not self.regex and not self.word_list) or (self.regex and self.word_list):
+            raise NightfallUserError("need either regex or word_list to build an ExclusionRule", 40001)
+
+    def as_dict(self):
```

Comment:
why use `as_dict` as opposed to just implementing `__dict__` so that the user can use the more pythonic `dict(inst)` as opposed to `inst.as_dict()`

Reply:
no reason, just me being silly